### ISSUE TITLE
Grub4dos/Grub4efi:  BUG in cat --replace command on NTFS volumes

### ISSUE THREAD

**Issue by steve6375** [chenall/grub4dos#361]
Bug in latest UDF-fixed grub4dos and also 2022-01-18 version.
Works fine in older versions - e.g. 2019, 2020-08-09.
Same bug in grub4efi 

```
!BAT

# fill (rd) with 0's , 0x82d0 is rd-base mem address, Fn24 is memset - fill memory  <addr> <string> <size> 
call Fn.24  0x60000 0x20  102400 > nul ;; map (md)0x300+200 (rd) > nul

errorcheck on
debug 3
debug msg=3

dd if=(rd)+1 of=()/AutoUnattend.xml
dd if=(bd)/My.xml of=()/AutoUnattend.xml
cat --locatei=LOADISO --replace=XYZX (hd0,0)/AutoUnattend.xml

cat --locatei=LOADISO (hd0,0)/AutoUnattend.xml && echo FILE NOT PATCHED!
```

![image](https://user-images.githubusercontent.com/1469350/178782376-ed332d84-74ad-416a-9ac4-6d6b78de40eb.png)
![image](https://user-images.githubusercontent.com/1469350/178784220-c4f0f9cd-24e7-423b-bf61-782a72ee0ede.png)

wrong place in file is patched!

![image](https://user-images.githubusercontent.com/1469350/178783129-d5d84b48-6b98-4fc9-a41e-efcc21e38dfb.png)

[testcat.zip](https://github.com/chenall/grub4dos/files/9104178/testcat.zip)

**Comment by steve6375** [comment]
**raw** cat  also fails in same way.

**Comment by steve6375** [comment]
Here is older grub4dos
![image](https://user-images.githubusercontent.com/1469350/178837417-436e1cf3-254c-4630-b77f-c55e2f3b8fd3.png)

Finds bytes at same position and replaces correctly.

**Comment by yaya2007** [comment]
Test with grub4dos-0.4.6a-2022-01-18, normal. The problem did not recur.

**Comment by steve6375** [comment]
![image](https://user-images.githubusercontent.com/1469350/178915847-bbef0e6c-1a28-4ee6-9610-4ff4effa834c.png)

Occurs for me  in VBox 5 and Lenovo IdeaPad 300 real system

File system is NTFS

**Comment by steve6375** [comment]
problem does not occur on FAT32 !!!
![image](https://user-images.githubusercontent.com/1469350/178921031-d9933cdb-5405-41fe-a30d-1b282ef153d2.png)

```
!BAT

# fill (rd) with 0's , 0x82d0 is rd-base mem address, Fn24 is memset - fill memory  <addr> <string> <size> 
call Fn.24  0x60000 0x20  102400 > nul ;; map (md)0x300+200 (rd) > nul

errorcheck on
debug 3
debug msg=3

echo FAT32
dd if=(rd)+1 of=(hd0,1)/test/AutoUnattend.xml
dd if=(bd)/My.xml of=(hd0,1)/test/AutoUnattend.xml
cat --locatei=LOADISO --replace=XYZX (hd0,1)/test/AutoUnattend.xml
cat --locatei=LOADISO (hd0,1)/test/AutoUnattend.xml && echo FILE NOT PATCHED!
echo
echo

echo NTFS
dd if=(rd)+1 of=()/AutoUnattend.xml
dd if=(bd)/My.xml of=()/AutoUnattend.xml
cat --locatei=LOADISO --replace=XYZX (hd0,0)/AutoUnattend.xml
cat --locatei=LOADISO (hd0,0)/AutoUnattend.xml && echo FILE NOT PATCHED!

```

**Comment by steve6375** [comment]
https://github.com/chenall/grub4dos/commit/edbca3ff07b7c455106471f1f4889838c61c9e9f#diff-97c5d5ae16bca1642c5cf4f6e32abb65c1e65211c16d2afbb89b00378294ecbc
![image](https://user-images.githubusercontent.com/1469350/179388590-5bf01829-2b3f-4698-8438-a6e0763277a3.png)

Possible these changes caused issue?
```
  Fixed write error when NTFS has offset.
  Fixed bug with dd function moving backwards in the same file.
```

**Comment by yaya2007** [comment]
yes. Related to this. The EFI was corrected last time, which implicated the BIOS. Now it is found that EFI's NTFS cache requires 4K alignment. Testing.

**Comment by yaya2007** [comment]
have a try。
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/9127359/grldr.rar.txt)
[BOOTX64.rar.txt](https://github.com/chenall/grub4dos/files/9127364/BOOTX64.rar.txt)

**Comment by steve6375** [comment]
They both seem to work OK - many thanks!
I tested for  redir bug #297  , dd bug #336  and this cat --replace bug #361.

**Comment by steve6375** [comment]
Problem with new grub4efi test version
Can no longer load .efi files using chainloader ???


e.g.
![image](https://user-images.githubusercontent.com/1469350/179397113-5c73ede8-0e49-41f7-83b0-5ac988867735.png)
![image](https://user-images.githubusercontent.com/1469350/179397243-d60ff18d-ff43-47fd-a4cf-0256962404ba.png)

**Comment by yaya2007** [comment]
Partition type is NTFS?

**Comment by steve6375** [comment]
yes NTFS.  FAT32 seems OK I think

**Comment by steve6375** [comment]
![image](https://user-images.githubusercontent.com/1469350/179400990-4d963ec7-8517-4c6a-add9-1392fbc0cf65.png)

**Comment by yaya2007** [comment]
UEFI shell does not support NTFS partitions.

First, execute in the fat partition:
 load /ntfs efi，
Then execute in NTFS partition:
 chainloader /grubfmx64 efi
 boot

**Comment by steve6375** [comment]
That works.
I didnt realise that chainloader command used the shell.
Why does chainloader need the shell and NTFS driver when grub4efi can read files on NTFS?
Can error message be improved? e.g. "file/volume not found".
Do any other commands require filesystem drivers/shell?

This means that we MUST have a FAT partition BUT I will work around this by making a floppy .ima image file with the ntfs.efi file inside and load the efi file after a map command.
```
#load filesystem drivers (for chainload *.efi and UEFI shell)
map --mem (bd)/%grub%/NTFS_EFI.ima.gz (5)
load (5)/NTFS.efi
map --unmap=5
```

**Comment by yaya2007** [comment]
Grub4dos itself has no disk drive. In BIOS mode, int13 function provided by firmware. In UEFI mode, through the shell function provided by the firmware. Usually, this shell does not contain NTFS drivers.

**Comment by a1ive** [comment]
> UEFI shell does not support NTFS partitions.
> 
> First, execute in the fat partition: load /ntfs efi， Then execute in NTFS partition: chainloader /grubfmx64 efi boot

启动 EFI 可执行程序不一定需要它在 FAT 分区上。你可以试一下 grub2，就可以启动任意分区下的 EFI 文件。
![1](https://user-images.githubusercontent.com/10670106/179513749-c888082f-1a45-4b7a-af48-377df7d26e19.png)
你只要在 grub 下预先为 EFI 文件分配好内存，把它复制过去，就可以用 LoadImage 加载内存中的 EFI 程序。

**Comment by yaya2007** [comment]
> 启动 EFI 可执行程序不一定需要它在 FAT 分区上。你可以试一下 grub2，就可以启动任意分区下的 EFI 文件。

下一个补丁会解决这个问题。

**Comment by steve6375** [comment]
FYI

map command wipes out NTFS driver load
NTFS.efi on FAT32 ptn
Ubuntu on NTFS ptn


```
load (hd0,1)/EFI/NTFS.efi
map /ubuntu.iso (0xff)
chainloader (0xff)
boot
```
>> FAIL
![image](https://user-images.githubusercontent.com/1469350/179705826-5c4a2d31-aad2-4eb2-b6d9-c9dbbd441e48.png)


```
map /ubuntu.iso (0xff)
load (hd0,1)/EFI/NTFS.efi
chainloader (0xff)
boot
```
>> BOOTS

![image](https://user-images.githubusercontent.com/1469350/179706031-12fc941a-9aa3-4dfc-98fe-7c6b40428479.png)
![image](https://user-images.githubusercontent.com/1469350/179706106-4cd1a914-314d-4bdb-a30a-dabebb9d62e4.png)

**Comment by yaya2007** [comment]
NTFS.efi应当放在FAT分区。你提供的磁盘样本，(hd0,0)是FAT分区，(hd0,1)是NTFS分区吧。

**Comment by steve6375** [comment]
no
hd0,0 = NTFS  - has Ubuntu iso
hd0,1 = FAT32  - has NTFS.efi
